src/Microsoft.Extensions.WebSockets/WebSocketConnection.cs

```diff
@@ -0,0 +1,391 @@
+        /// </summary>
+        /// <param name="inbound">A <see cref="IReadableChannel"/> from which frames will be read when receiving.</param>
+        /// <param name="outbound">A <see cref="IWritableChannel"/> to which frame will be written when sending.</param>
+        /// <param name="subProtocol">The subprotocol (if any) used by this connection. Has no effect on behavior, just used to fill the <see cref="SubProtocol"/> property for callers to read later.</param>
+        /// <param name="masked">A boolean indicating if frames sent from this socket should be masked (the masking key is automatically generated)</param>
+        public WebSocketConnection(IReadableChannel inbound, IWritableChannel outbound, bool masked, string subProtocol)
+        {
+            SubProtocol = subProtocol;
+            _inbound = inbound;
+            _outbound = outbound;
+
+            if (masked)
+            {
+                _maskingKey = new byte[4];
+                _randomMask = true;
+            }
+        }
+
+        /// <summary>
+        /// Constructs a new, fixed masking-key, <see cref="WebSocketConnection"/> from an <see cref="IReadableChannel"/> and an <see cref="IWritableChannel"/> that represents an established WebSocket connection (i.e. after handshaking)
+        /// </summary>
+        /// <param name="inbound">A <see cref="IReadableChannel"/> from which frames will be read when receiving.</param>
+        /// <param name="outbound">A <see cref="IWritableChannel"/> to which frame will be written when sending.</param>
+        /// <param name="subProtocol">The subprotocol (if any) used by this connection. Has no effect on behavior, just used to fill the <see cref="SubProtocol"/> property for callers to read later.</param>
+        /// <param name="fixedMaskingKey">The masking key to use for the connection. Must be exactly 4-bytes long. This is ONLY recommended for testing and development purposes.</param>
+        public WebSocketConnection(IReadableChannel inbound, IWritableChannel outbound, string subProtocol, byte[] fixedMaskingKey)
+        {
+            SubProtocol = subProtocol;
+            _inbound = inbound;
+            _outbound = outbound;
+            _maskingKey = fixedMaskingKey;
+        }
+
+        /// <summary>
+        /// Receives the next available frame.
+        /// </summary>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> that indicates when/if the receive is cancelled.</param>
+        /// <returns>A <see cref="Task{WebSocketMessage}"/> that completes with the message when it has been received.</returns>
+        // TODO: ValueTask?
+        public async Task<WebSocketFrame> ReceiveAsync(CancellationToken cancellationToken)
+        {
+            // Mark the previous frame buffer (if any) as consumed.
+            if (_lastPayloadEnd != null)
+            {
+                _inbound.Advance(_lastPayloadEnd.Value);
```

Comment:
odd timing; could this be done better by `Preserve()`-ing the existing payload and advancing the stream immediately? rather than leaving it dangling? admittedly, that then places a requirement to dispose on the consumer...